### PR TITLE
Rewrite builder and channel manager

### PR DESCRIPTION
**Pull request by Ronserruya**
* Main purpose:
 Rewrite the sdk's channel manager after feedback from different teams, and make it easier to support direct access to the builder object

* Technical description:

- KinAccount and Builder are now exposed to the user
- Extra checks like checking if all channels are created before starting the sdk were removed
- When creating channels, we do some checks to see if we can pass it as a no-op
- Builder was changed to support changing the channel after the builder was created
- Sign method was removed from the Kin-sdk builder, so using 'sign()' no longer updates the sequence inexplicitly 
- Complete rewrite of the channel manager to use a custom pool/queue object
- The channel manager now only deals with channels, not building/sending transactions
- Removed the kin. Transaction object since now the builders are exposed
- Added get_status method to KinAccount to show info on the account and channels
- Fixed typos and added TODOs :)

* Tests
Tests are failing for python 2 because of an issue with kin-base, this will be fixed later, we should also send a PR to the original py-stellar-base

* Documentation (Source/readme.md)
Readme was weekly updated since we will a nicer documentation soon anyway

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -200,23 +200,25 @@ tx_hash = account.send_kin('destination', 1000, fee=100, memo_text='order123')
 ```
 
 ### Build/Submit transactions
-While the previous methods build and send the transaction for you, there is another way to send transactions with two steps
+While the previous methods build and send the transaction for you, there is another way to send transactions
 
 Step 1: Build the transaction
 ```python
-tx = account.build_send_kin('destination',1000, fee=100, memo_text='order123')
+builder = account.build_send_kin('destination',1000, fee=100, memo_text='order123')
```
Comment (doodyparizada) [inline README.md]: pep8 also in documentation: missing a space `'destination', 1000`

File: README.md [hunk 2]
```diff
@@ -200,23 +200,25 @@ tx_hash = account.send_kin('destination', 1000, fee=100, memo_text='order123')
...
 ```
-Step 2: Submit the transaction
+Step 2: Update the transaction
 ```python
-tx_hash = account.submit_transaction(tx)
+# do whatever you want with the builder
+with account.channel_manager.get_channel() as channel:
+    builder.set_channel(channel)
+    builder.sign(channel)
+    # If you used additional channels apart from your main account,
```
Comment (doodyparizada) [inline README.md]: `As this example uses additional channels...`

File: kin/blockchain/builder.py [hunk 3]
```diff
@@ -3,33 +3,31 @@
...
+        :param str network_name: The name of the network
+        :param Kin.Horizon horizon: The horizon instance to use
+        :param int fee: Fee for the transaction
+        :param str secret: The seed to be used
+        """
 
-        # call baseclass constructor to init base class variables
-        super(Builder, self).__init__(secret=secret, address=address, sequence=1, fee=fee)
+        # call base class constructor to init base class variables
+        # sequence is one since its get overridden later
```
Comment (doodyparizada) [inline kin/blockchain/builder.py]: > since it gets

Comment (doodyparizada) [reply kin/blockchain/builder.py]: BTW, why pass this as a magic number? why not send `None` or even 0 or not pass it at all?

File: kin/blockchain/builder.py [hunk 4]
```diff
@@ -3,33 +3,31 @@
...
+        super(Builder, self).__init__(secret=secret, sequence=1, fee=fee)
 
         # custom overrides
 
-        self.network = network
+        self.network = network_name
```
Comment (doodyparizada) [inline kin/blockchain/builder.py]: why not keep the same name `self.network_name = network_name`

File: kin/blockchain/builder.py [hunk 5]
```diff
@@ -40,9 +38,16 @@ def clear(self):
...
-        return self.horizon.account(self.address).get('sequence')
+    def update_sequence(self):
+        """
+        Update the builder with the *current* sequence of the account
+        # TODO: kin-base builder increments this value by 1 when building a tx.
+        #       Remove this functionality from py-stellar-base and change this method set the current sequence+1
+        """
+
+        # TODO: kin-base checks for 'not sequence' to find if there is no sequence, therefore
+        # Sequence of 0 fails, write it as a str for now and fix in kin-base later
```
Comment (doodyparizada) [inline kin/blockchain/builder.py]: aren't you using a default of `sequence = 1` though? Why do you need this hack then?

File: kin/blockchain/channel_manager.py [hunk 6]
```diff
@@ -1,77 +1,143 @@
...
-            builder.sign()  # always sign with a channel key
-            if source:
-                builder.sign(secret=self.base_key)  # sign with the base key if needed
-        except:
-            # If something fails when building the tx, clear and return the builder to queue.
-            builder.clear()
-            self.channel_builders.put(builder,timeout=CHANNEL_PUT_TIMEOUT)
-            raise
-        return builder
+            self.channel_pool.put(channel, timeout=timeout)
```
Comment (doodyparizada) [inline kin/blockchain/channel_manager.py]: Why would the queue ever get full? You are managing your own pool.

File: kin/blockchain/channel_manager.py [hunk 7]
```diff
@@ -1,77 +1,143 @@
...
+        self.channel_pool = ChannelPool(channel_seeds)
 
-        :param str memo_text: (optional) a text to add as transaction memo.
+    @contextmanager
+    def get_channel(self, timeout=CHANNEL_GET_TIMEOUT):
+        """
+        Get an available channel
+        :param float timeout: (Optional) How long to wait before raising an exception
+        :return a free channel seed
+        :rtype str
```
Comment (doodyparizada) [inline kin/blockchain/channel_manager.py]: sometimes you're using `return` and sometimes `rtype`

File: kin/blockchain/channel_manager.py [hunk 8]
```diff
@@ -1,77 +1,143 @@
...
+
+    def get_status(self, verbose=False):
+        """
+        Return the current status of the channel manager
+        :param bool verbose: Include all channel seeds and their statuses in the response
+        :return: dict
+        """
+        free_channels = len(self.channel_pool.get_free_channels())
+        status = {
+            'total_channels': len(self.channel_pool.queue),
```
Comment (doodyparizada) [inline kin/blockchain/channel_manager.py]: if you're pre calculating `free_channels` to keep DRY, you might as well do the same to `total_channels`

File: kin/blockchain/channel_manager.py [hunk 9]
```diff
@@ -1,77 +1,143 @@
...
+        # Base queue checks if the queue is not empty by checking the length of the queue (_qsize() != 0)
+        # We need to check it by checking how many channels are free
+        return len(self.get_free_channels())
+
+    def get_free_channels(self):
+        """
+        Get a list of channels with "FREE" status
+        :rtype list[str]
+        """
+        return list(filter(lambda key: self.queue[key] == ChannelStatuses.FREE, self.queue.keys()))
```
Comment (doodyparizada) [inline kin/blockchain/channel_manager.py]: I think list comprehension is more readable: 
```python
[channel for channel, status in self.queue.items() if status == ChannelStatuses.FREE]
```

File: kin/blockchain/builder.py (hunk 3 above)
Comment (Ronserruya) [reply kin/blockchain/builder.py]: kin-base will query horizon if you don't give a sequence number. Until I change this I chose to pass it like this for now

File: kin/blockchain/builder.py (hunk 4 above)
Comment (Ronserruya) [reply kin/blockchain/builder.py]: "network" is used in kin-base, so I cant only change it here

File: kin/blockchain/builder.py (hunk 5 above)
Comment (Ronserruya) [reply kin/blockchain/builder.py]: If you use the root account, its first sequence will be 0. kin-base will then do:
```if not self.sequence:
            raise SequenceError('No sequence is present, maybe not funded?')
```
So a sequence if 0 will raise an error. That's why I pass it a '0'

File: kin/blockchain/channel_manager.py (hunk 6 above)
Comment (Ronserruya) [reply kin/blockchain/channel_manager.py]: There shouldn't be a case where it gets full, however, if someone tries to directly call channel_manager.put he will get a clear error

**Pull request merged**